### PR TITLE
[CALCITE-4411] RelNode to SQL lost the distinct when doing window aggregation (Jiatao Tao)

### PR DESCRIPTION
**Pull request by Aaaaaaron**

### CODE REVIEW

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java [hunk 1]
```diff
@@ -3008,6 +3008,12 @@ private void checkLiteral2(String expression, String expected) {
             + "OVER (PARTITION BY \"hire_date\" ORDER BY \"employee_id\"), \"hire_date\"\n"
             + "FROM \"foodmart\".\"employee\"\n"
             + "GROUP BY \"hire_date\", \"employee_id\"";
+    String query7 = "SELECT "
+        + "count(distinct \"employee_id\") over (order by \"hire_date\") FROM \"employee\"";
+    String expected7 = "SELECT "
+        + "COUNT(DISTINCT \"employee_id\") "
+        + "OVER (ORDER BY \"hire_date\" RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS \"$0\""
+        + "\nFROM \"foodmart\".\"employee\"";
```
Comment (danny0405) [inline core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:3016]: `RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW` should be the default and not print out.

File: core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java [hunk 2]
```diff
@@ -1007,19 +1007,27 @@ private SqlCall toSql(RexProgram program, RexOver rexOver) {
...
     }
 
     private SqlCall createOverCall(SqlAggFunction op, List<SqlNode> operands,
-        SqlWindow window) {
+        SqlWindow window, boolean isDistinct) {
       if (op instanceof SqlSumEmptyIsZeroAggFunction) {
         // Rewrite "SUM0(x) OVER w" to "COALESCE(SUM(x) OVER w, 0)"
         final SqlCall node =
-            createOverCall(SqlStdOperatorTable.SUM, operands, window);
+            createOverCall(SqlStdOperatorTable.SUM, operands, window, isDistinct);
```
Comment (zabetak) [inline core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java:1018]: Can we enter here with `isDistinct == true`? If yes can you also add a testcase?

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java (hunk 1 above)
Comment (Aaaaaaron) [reply core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:3016]: Sorry don't get your opinion @zabetak , if you just run the test without my modification, it'll fail, the distinct is lost, more details can be seen in JIRS

File: core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java (hunk 2 above)
Comment (Aaaaaaron) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java:1018]: Sorry?

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java (hunk 1 above)
Comment (Aaaaaaron) [reply core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:3016]: @danny0405 Hi Danny, seems rel to SQL will print these, if I remove them, the test failed, maybe we can open a new JIRA to not print out these in RelToSql.

File: core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java (hunk 2 above)
Comment (Aaaaaaron) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java:1018]: Seems can enter, I test "SELECT sum(distinct employee_id) over (order by hire_date) FROM employee" can enter, but I can not write a SQL with SUM0, do you have any advice about this, or I can add the SUM DISTINCT query?

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java (hunk 1 above)
Comment (danny0405) [reply core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:3016]: We can fix it in separate issue and PR.

Comment (Aaaaaaron) [reply core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:3016]: ok

File: core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java (hunk 2 above)
Comment (zabetak) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java:1018]: If I remember well `SUM0` appears when the column is nullable. Did you try an example with a nullable column?

Comment (Aaaaaaron) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java:1018]: ok, I'll take a try

**Review by chunweilei — COMMENTED** [review]
LGTM

**Review by chunweilei — APPROVED** [review]
+1

**Comment by chunweilei** [comment]
One minor comments: use `loses` rather than `lost` in the commit message.

**Comment by zabetak** [comment]
I think it will be nice to have an extra test (as mentioned above) if that's possible.

**Comment by Aaaaaaron** [comment]
> I think it will be nice to have an extra test (as mentioned above) if that's possible.

Hi @zabetak I've added a query(query8), can you take a look?

**Comment by Aaaaaaron** [comment]
> One minor comment: use `loses` rather than `lost` in the commit message.

Revised.

**Comment by zabetak** [comment]
@Aaaaaaron I am testing locally, if all goes well I will merge this right after!

**Comment by Aaaaaaron** [comment]
> @Aaaaaaron I am testing locally, if all goes well I will merge this right after!

@zabetak Thanks for your patient, really appreciate it!

**Pull request closed without merging**